### PR TITLE
HiFiC modules

### PR DESCRIPTION
**Pull request by 0x00b1**
Implements the following modules from Mentzer, et al. (2020)

- HiFiCDiscriminator
- HiFiCEncoder
- HiFiCGenerator

```bibtex
@misc{mentzer2020highfidelity,
      title={High-Fidelity Generative Image Compression}, 
      author={Fabian Mentzer and George Toderici and Michael Tschannen and Eirikur Agustsson},
      year={2020},
      eprint={2006.09965},
      archivePrefix={arXiv},
      primaryClass={eess.IV}
}
```

Originally implemented in TensorFlow Compression (TFC) by the author (@relational).

### CODE REVIEW

**Comment by 0x00b1** [comment]
I'm not sure how to deal with the MyPy `kwargs` issue without inlining. @mmuckley Any suggestions?

**Comment by mmuckley** [comment]
Hello @0x00b1, I'm not sure it's that easy as `**kwargs` could have essentially any type. That being said I think I agree with `mypy` - `**kwargs` can pass on essentially anything and as such can be a breeding ground for bugs. My suggestion would be to just fully specify all `kwargs` and let users subclass and extend the models when needed to pass new parameters.

Note that this is primarily due to the code being in the core repository - if it were under `projects` then I would say we can adopt a soft linting standard and you could just disable `mypy` for your folder.

File: neuralcompression/models/_hific/_channel_norm_2d.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import torch
+import torch.nn
+
+
+def _channel_norm_2d(
+    input_channels, momentum=0.1, affine=True, track_running_stats=False, **kwargs
```
Comment (mmuckley) [inline neuralcompression/models/_hific/_channel_norm_2d.py:6]: `**kwargs` unused here.

File: neuralcompression/models/_hific/_channel_norm_2d.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        input_channels,
+        momentum=momentum,
+        affine=affine,
+        track_running_stats=track_running_stats,
+    )
+
+
+class ChannelNorm2D(torch.nn.Module):
+    def __init__(
+        self, input_channels, momentum=0.1, epsilon=1e-3, affine=True, **kwargs
```
Comment (mmuckley) [inline neuralcompression/models/_hific/_channel_norm_2d.py:18]: `**kwargs` unused here.

File: neuralcompression/models/_hific/_hific_discriminator.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+        x = self.activation(self.conv_1(x))
+
+        x = self.activation(self.conv_2(x))
+        x = self.activation(self.conv_3(x))
+        x = self.activation(self.conv_4(x))
+
+        predictions = self.predictions(x).view(-1, 1)
+
+        return torch.sigmoid(predictions), predictions
```
Comment (mmuckley) [inline neuralcompression/models/_hific/_hific_discriminator.py:70]: Proposing a `mypy`-friendly version.
```suggestion
        self.input_sequence = torch.nn.Sequential(
            torch.nn.LeakyReLU(0.2),
            torch.nn.Conv2d(
                self.features, 12, (3, 3), padding=1, padding_mode="reflect"
            ),
            torch.nn.Upsample(None, 16, "nearest"),
        )

        filters = (64, 128, 256, 512)

        conv_sequence = []
        old_size = self.image_dimensions[0] + 12
        for current_filter in filters:
            conv_sequence.append(
                torch.nn.utils.spectral_norm(
                    torch.nn.Conv2d(
                        old_size,
                        current_filter,
                        (4, 4),
                        padding=1,
                        padding_mode="reflect",
                        stride=2,
                    )
                )
            )
            old_size = current_filter

        conv_sequence.append(torch.nn.Conv2d(filters[3], 1, (1, 1), (1, 1)))

        self.conv_sequence = torch.nn.Sequential(*conv_sequence)

    def forward(
        self, x: torch.Tensor, y: torch.Tensor
    ) -> typing.Tuple[torch.Tensor, torch.Tensor]:
        predictions = self.conv_sequence(
            torch.cat((x, self.input_sequence(y)), 1)
        ).view(-1, 1)

        return torch.sigmoid(predictions), predictions
```

File: neuralcompression/models/_hific/_hific_encoder.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
+import typing
+
+import torch.nn
+
+from ._channel_norm_2d import _channel_norm_2d
+
+
+class HiFiCEncoder(torch.nn.Module):
```
Comment (mmuckley) [inline neuralcompression/models/_hific/_hific_encoder.py:8]: Same as above - I think we can avoid any use of `kwargs` with loops and `torch.nn.Sequential`.

File: neuralcompression/models/_hific/_hific_generator.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
+import typing
+
+import torch
+import torch.nn
+import torch.nn.functional
+
+from ._channel_norm_2d import _channel_norm_2d
+
+
+class _ResidualBlock(torch.nn.Module):
```
Comment (mmuckley) [inline neuralcompression/models/_hific/_hific_generator.py:10]: See above on `Sequential`.

File: neuralcompression/models/_hific/_hific_generator.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+        features = self.pad(features)
+
+        features = self.conv_b(features)
+        features = self.norm_b(features)
+
+        return torch.add(features, identity_map)
+
+
+class HiFiCGenerator(torch.nn.Module):
```
Comment (mmuckley) [inline neuralcompression/models/_hific/_hific_generator.py:57]: See above on `Sequential`

**Review by mmuckley — COMMENTED** [review]
Making some recommendations to pass the lints easier - I think it's great to use `torch.nn.Sequential` and loops to organize things. There shouldn't be any compute cost - these are just in `__init__` and the loops are tiny.

**Comment by 0x00b1** [comment]
@mmuckley Thanks! I love the changes.

File: neuralcompression/models/_hific/__init__.py [hunk 7]
```diff
@@ -0,0 +1,3 @@
+from ._hific_discriminator import HiFiCDiscriminator
```
Comment (mmuckley) [inline neuralcompression/models/_hific/__init__.py:1]: Please add a copyright header.
```
"""
Copyright (c) Facebook, Inc. and its affiliates.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
"""
```

File: neuralcompression/models/_hific/_channel_norm_2d.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
+import torch
```
Comment (mmuckley) [inline neuralcompression/models/_hific/_channel_norm_2d.py:1]: Please add a copyright header.
```
"""
Copyright (c) Facebook, Inc. and its affiliates.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
"""
```

File: neuralcompression/models/_hific/_hific_discriminator.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
+import typing
```
Comment (mmuckley) [inline neuralcompression/models/_hific/_hific_discriminator.py:1]: Please add a copyright header.
```
"""
Copyright (c) Facebook, Inc. and its affiliates.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
"""
```

File: neuralcompression/models/_hific/_hific_encoder.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
+import typing
```
Comment (mmuckley) [inline neuralcompression/models/_hific/_hific_encoder.py:1]: Please add a copyright header.
```
"""
Copyright (c) Facebook, Inc. and its affiliates.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
"""
```

File: neuralcompression/models/_hific/_hific_generator.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
+import typing
```
Comment (mmuckley) [inline neuralcompression/models/_hific/_hific_generator.py:1]: Please add a copyright header.
```
"""
Copyright (c) Facebook, Inc. and its affiliates.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
"""
```

File: tests/models/_hific/test_hific_discriminator.py [hunk 12]
```diff
@@ -0,0 +1,16 @@
+import torch
```
Comment (mmuckley) [inline tests/models/_hific/test_hific_discriminator.py:1]: Please add a copyright header.
```
"""
Copyright (c) Facebook, Inc. and its affiliates.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
"""
```

File: tests/models/_hific/test_hific_generator.py [hunk 13]
```diff
@@ -0,0 +1,14 @@
+import torch
```
Comment (mmuckley) [inline tests/models/_hific/test_hific_generator.py:1]: Please add a copyright header.
```
"""
Copyright (c) Facebook, Inc. and its affiliates.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
"""
```

**Review by mmuckley — CHANGES_REQUESTED** [review]

**Review by mmuckley — APPROVED** [review]
LGTM.

**Pull request merged**